python/ray/train/v2/api/train_fn_utils.py

```diff
@@ -1,36 +1,206 @@
 from typing import TYPE_CHECKING, Any, Dict, Optional
+from ray.util.annotations import PublicAPI
 
 from ray.train import Checkpoint
-from ray.train._internal import session
-from ray.train.context import get_context as get_context_v1
-from ray.train.v2._internal.execution.context import TrainContext, get_train_context
-from ray.train.v2._internal.util import _copy_doc
+from ray.train.v2._internal.execution.context import get_train_context
+from ray.train.v2.api.context import TrainContext
 
 if TYPE_CHECKING:
     from ray.data import DataIterator
 
 
-@_copy_doc(session.report)
+@PublicAPI(stability="stable")
 def report(
     metrics: Dict[str, Any],
     checkpoint: Optional[Checkpoint] = None,
     checkpoint_dir_name: Optional[str] = None,
 ):
+    """Report metrics and optionally save a checkpoint.
+
+    If a checkpoint is provided, it will be
+    :ref:`persisted to storage <persistent-storage-guide>`.
+
+    If this is called in multiple distributed training workers:
+
+    - Only the metrics reported by the rank 0 worker will be tracked by Ray Train.
+      See :ref:`the metrics logging guide <train-monitoring-and-logging>`.
+    - A checkpoint will be registered as long as one or more workers reports
+      checkpoint that is not None.
+      See the :ref:`checkpointing guide <train-dl-saving-checkpoints>`.
+    - Checkpoints from multiple workers will be merged into one directory
+      in persistent storage.
+      See :ref:`the distributed checkpointing guide <train-distributed-checkpointing>`.
+
+    .. note::
+
+        Each invocation of this method will automatically increment the underlying
+        ``training_iteration`` number. The physical meaning of this "iteration" is
+        defined by user depending on how often they call ``report``.
+        It does not necessarily map to one epoch.
+
+    .. warning::
+
+        All workers must call `ray.train.report` the same number of times
+        so that Ray Train can properly synchronize the training state across
+        workers. Otherwise, your training will hang.
```

Comment:
```suggestion
        All workers must call `ray.train.report` the same number of times
        so that Ray Train can properly synchronize the training state across
        workers. This method acts as a barrier across all workers, so be sure
        that every worker reaches this method.
```